File: DSA/3. Expensive Expeditions/Part 3 (Medium)/Python/judge.py
from collections import Counter
from solution import Solution

class Judge:
    def run_tests(self):
        test_cases = [
            "h-H-+e-e-+lll-l-o-o",
            "Goodbye-------+oodmorning!",
            "Foo_----_-B-+a-+r-+",
        ]
        for test_case in test_cases:
            self.test(test_case, self.test_case_solution_output(test_case))

    def test(self, p, expected):
        solution = Solution()
        result = solution.decode_message(p)
        passed = result == expected
        print(f"Message: {p}")
        print(f"Expected: {expected}")
        print(f"Your Output: {result}")
        print("Test " + ("PASSED" if passed else "FAILED"))
        print("-----------")

    def test_case_solution_output(self, m):
        stack = []
        prev = None
        for char in m:
            if char == '-':
                if stack:
                    prev = stack.pop()
            elif char == '+':
                if prev:
                    stack.append(prev)
            else:
                stack.append(char)
        return ''.join(stack)

File: DSA/3. Expensive Expeditions/Part 3 (Medium)/JavaScript/judge.js
const Solution = require('./solution');

class Judge {
    runTests() {
        const testCases = [
            "h-H-+e-e-+lll-l-o-o",
            "Goodbye-------+oodmorning!",
            "Foo_----_-B-+a-+r-+",
        ];

        testCases.forEach(testCase => {
            this.test(testCase, this.testCaseSolutionOutput(testCase));
        });
    }

    test(p, expected) {
        const solution = new Solution();
        const result = solution.decodeMessage(p);
        const passed = result === expected;
        console.log(`Message: ${p}`);
        console.log(`Expected: ${expected}`);
        console.log(`Your Output: ${result}`);
        console.log(passed ? "Test PASSED" : "Test FAILED");
        console.log("-----------");
    }

    testCaseSolutionOutput(p) {
        const stack = [];
        let prev = null;
        for (const char of p) {
            if (char === '-') {
                if (stack.length > 0) {
                    prev = stack.pop();
                }
            } else if (char === '+') {
                if (prev !== null) {
                    stack.push(prev);
                }
            } else {
                stack.push(char);
            }
        }
        return stack.join('');
    }
}

module.exports = Judge;